sirepo/package_data/static/js/sirepo.js

```diff
@@ -1038,6 +1041,345 @@ SIREPO.app.factory('timeService', function() {
+    const _e_authError = 'authError';
+    // incorrect credentials from sbatchLogin
+    const _e_authInvalid = 'authInvalid';
+    // logged out reply from sbatchLoginStatus
+    const _e_authMissing = 'authMissing';
+    // logged in reply from sbatchLoginStatus or ready from sbatchLogin
+    const _e_authSuccess = 'authSuccess';
+    // cancel pressed by user before credential input
+    const _e_credsCancel = 'credsCancel';
+    // submit pressed on creds form
+    const _e_credsConfirm = 'credsConfirm';
+    // login button pressed to show creds form
+    const _e_loginClicked = 'loginClicked';
+    // change jobRunMode != sbatch
+    const _e_needNo = 'needNo';
+    // change jobRunMode == sbatch
+    const _e_needYes = 'needYes';
+    // models have been unloaded
+    const _e_unloaded = 'unloaded';
+
+    const _REASON_TO_EVENTS = {
+        "general-connection-error": _e_authError,
+        "invalid-creds": _e_authInvalid,
+        "no-creds": _e_authMissing,
+    };
+
+    const _TRANSITIONS = {
+        [_s_auth]: {
+            [_e_authError]: _s_creds,
+            [_e_authInvalid]: _s_creds,
+            [_e_authMissing]: _s_creds,
+            [_e_authSuccess]: _s_ok,
+            [_e_credsCancel]: _s_idle,
+            [_e_needYes]: _s_auth,
+            [_e_unloaded]: _s_initial,
+        },
+        [_s_creds]: {
+            [_e_authMissing]: _s_creds,
+            [_e_credsCancel]: _s_idle,
+            [_e_credsConfirm]: _s_auth,
+            [_e_needYes]: _s_creds,
+            [_e_unloaded]: _s_initial,
+        },
+        [_s_idle]: {
+            [_e_loginClicked]: _s_creds,
+            [_e_needNo]: _s_notNeeded,
+            [_e_needYes]: _s_status,
+            [_e_authMissing]: _s_idle,
+            [_e_unloaded]: _s_initial,
+        },
+        [_s_initial]: {
+            [_e_needNo]: _s_notNeeded,
+            [_e_needYes]: _s_status,
+            [_e_authMissing]: _s_idle,
+            [_e_unloaded]: _s_initial,
+        },
+        [_s_notNeeded]: {
+            [_e_authMissing]: _s_idle,
+            [_e_needNo]: _s_notNeeded,
+            [_e_needYes]: _s_status,
+            [_e_unloaded]: _s_initial,
+        },
+        [_s_ok]: {
+            [_e_authMissing]: _s_idle,
+            [_e_needYes]: _s_ok,
+            [_e_unloaded]: _s_initial,
+        },
+        [_s_status]: {
+            [_e_authError]: _s_idle,
+            [_e_authMissing]: _s_idle,
+            [_e_authSuccess]: _s_ok,
+            [_e_needYes]: _s_status,
+            [_e_unloaded]: _s_initial,
+        },
+    };
+
+    const _STATES = new Set(Object.keys(_TRANSITIONS));
+
+    const _EVENTS =  new Set(Object.values(_TRANSITIONS).map((x) => {return Object.keys(x);}).flat(1));
+
+    let _state = _s_initial;
+
+    const _STATE_QUERIES = {
+        isLoggedIn: (state) => {
+            return state === _s_ok;
+        },
+        showLogin: (state) => {
+            return [_s_creds, _s_idle].includes(state);
```

Comment:
Sets are faster ([sort of](https://stackoverflow.com/questions/39007637/javascript-set-vs-array-performance)) and more logical for the general problem. My trade off is often the amount of distracting code. For the queries, it would be distracting to create a set inline and it would be slower than includes. For _EVENTS and _STATES, Set creation happens at initialization, and has() is faster and less text to give the same meaning.